pkg/await/states/types.go

```diff
@@ -0,0 +1,80 @@
+// Copyright 2016-2019, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package states
+
+import (
+	"fmt"
+
+	"github.com/pulumi/pulumi-kubernetes/pkg/logging"
+)
+
+// Result specifies the result of a Condition applied to an input object.
+type Result struct {
+	Ok          bool            // True if the Condition is true, false otherwise.
+	Description string          // A human-readable description of the associated Condition.
+	Message     logging.Message // The message to be logged after evaluating the Condition.
+}
+
+func (r Result) String() string {
+	s := fmt.Sprintf("[%t] %s", r.Ok, r.Description)
+
+	if !r.Message.Empty() {
+		s = fmt.Sprintf("%s -- %s", s, r.Message)
+	}
+
+	return s
+}
+
+// Condition is a function that checks a state and returns a Result.
+type Condition func(s interface{}) Result
+
+// StateChecker holds the data required to generically implement await logic.
```

Comment:
Might be nice to make this clearer, too. `StateChecker` checks that a `Pod` has a set of `v1.PodConditions` that match the desired state.

Reply:
Again, `StateChecker` isn't specific to `Pod`; it will be used for every resource type once I'm done with refactoring.